spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/FeignClientFactoryBean.java

```diff
@@ -91,40 +65,67 @@ public void afterPropertiesSet() throws Exception {
 		}
 	}
 
+	@Override
+	public void setApplicationContext(ApplicationContext context) throws BeansException {
+		this.context = context;
+	}
+
 	protected Feign.Builder feign() {
+		// @formatter:off
 		Feign.Builder builder = Feign.builder()
 				// required values
-				.logger(this.logger).encoder(this.encoder).decoder(this.decoder)
-				.contract(this.contract);
+				.logger(get(Logger.class))
+				.encoder(get(Encoder.class))
+				.decoder(get(Decoder.class))
+				.contract(get(Contract.class));
+		// @formatter:on
 
 		// optional values
-		if (this.logLevel != null) {
-			builder.logLevel(this.logLevel);
+		Logger.Level level = getOptional(Logger.Level.class);
+		if (level != null) {
+			builder.logLevel(level);
 		}
-		if (this.retryer != null) {
-			builder.retryer(this.retryer);
+		Retryer retryer = getOptional(Retryer.class);
+		if (retryer != null) {
+			builder.retryer(retryer);
 		}
-		if (this.errorDecoder != null) {
-			builder.errorDecoder(this.errorDecoder);
+		ErrorDecoder errorDecoder = getOptional(ErrorDecoder.class);
+		if (errorDecoder != null) {
+			builder.errorDecoder(errorDecoder);
 		}
-		if (this.options != null) {
-			builder.options(this.options);
+		Request.Options options = getOptional(Request.Options.class);
+		if (options != null) {
+			builder.options(options);
 		}
-		if (this.requestInterceptors != null) {
-			builder.requestInterceptors(this.requestInterceptors);
+		Map<String, RequestInterceptor> requestInterceptors = this.context.getBeansOfType(RequestInterceptor.class);
+		if (requestInterceptors != null) {
+			builder.requestInterceptors(requestInterceptors.values());
 		}
 
 		return builder;
 	}
 
+	protected <T> T get(Class<T> type) {
+		return this.context.getBean(type);
+	}
+
+	protected <T> T getOptional(Class<T> type) {
+		try {
+			return this.context.getBean(type);
+		} catch (NoSuchBeanDefinitionException e) {
+			//ignore
+		}
+		return null;
+	}
+
 	protected <T> T loadBalance(Feign.Builder builder, Class<T> type, String schemeName) {
 		builder.logger(new Slf4jLogger(type)); // TODO: how to have choice here?
-		if (this.ribbonClient != null) {
-			return builder.client(this.ribbonClient).target(type, schemeName);
-		}
-		else {
-			return builder.target(LoadBalancingTarget.create(type, schemeName));
+		Client client = getOptional(Client.class);
+		if (client != null) {
+			return builder.client(client).target(type, schemeName);
 		}
+
+		throw new IllegalStateException("No Feign Client for loadBalancing defined. Did you forget to include spring-cloud-starter-ribbon?");
```

Comment:
`builder.target(LoadBalancingTarget.create(type, schemeName))` would happen even if ribbon was not loaded, this is a problem too.